File: scala/parsers/src/main/scala/jp/co/teruuu/parser/common/Pair2.scala
package jp.co.teruuu.parser.common

class Pair2[T,U](lps: Parser[T], rps: Parser[U]) extends Parser[(T,U)] {
  override def parse(input: String): ParseResult[(T, U)] = {
    lps.parse(input) match {
      case lResult: ParseOk[T] => {
        rps.parse(lResult.next) match {
          case rResult: ParseOk[U] => ParseOk((lResult.value, rResult.value), rResult.next)
          case rResult: ParseNg[_] => ParseNg(rResult.message, rResult.next)
        }
      }
      case lResult: ParseNg[_] => ParseNg(lResult.message, lResult.next)
    }
  }
}

File: java/parsers_java8/src/main/java/jp/co/teruuu/parser/common/Pair2.java
package jp.co.teruuu.parser.common;

import jp.co.teruuu.parser.common.type.Tuple2;

public class Pair2<X, Y> implements Parser<Tuple2<X, Y>> {
    private Parser<X> lhs;
    private Parser<Y> rhs;
    public Pair2(Parser<X> lhs, Parser<Y> rhs) {
        this.lhs = lhs;
        this.rhs = rhs;
    }

    @Override
    public ParseResult<Tuple2<X, Y>> parse(String input) {
        ParseResult<X> lresult = lhs.parse(input);

        if(lresult instanceof ParseResult.Success<?>) {
            X value1 = ((ParseResult.Success<X>)lresult).value;
            String next1 = ((ParseResult.Success<X>)lresult).next;
            ParseResult<Y> rresult = rhs.parse(next1);

            if(rresult instanceof ParseResult.Success<?>) {
                Y value2 = ((ParseResult.Success<Y>)rresult).value;
                String next2 = ((ParseResult.Success<Y>) rresult).next;
                return new ParseResult.Success(new Tuple2(value1, value2), next2);
            } else {
                return (ParseResult.Failure) rresult;
            }
        } else {
            return (ParseResult.Failure)lresult;
        }
    }
}